compiler/rustc_mir_transform/src/lint_tail_expr_drop_order.rs

```diff
@@ -0,0 +1,382 @@
+                tcx.hir().body_owned_by(def_id).id().hir_id
+            ),
+            (lint::Level::Allow, _)
+        )
+    {
+        return;
+    }
+    // We are using blocks to identify locals with the same scope targeted by backwards-incompatible drops (BID)
+    // because they tend to be scheduled in the same drop ladder block.
+    let mut bid_per_block = IndexMap::default();
+    let mut bid_places = UnordSet::new();
+    for (block, data) in body.basic_blocks.iter_enumerated() {
+        for (statement_index, stmt) in data.statements.iter().enumerate() {
+            if let StatementKind::BackwardIncompatibleDropHint { place, reason: _ } = &stmt.kind {
+                bid_per_block
+                    .entry(block)
+                    .or_insert(vec![])
+                    .push((Location { block, statement_index }, &**place));
+                bid_places.insert(&**place);
+            }
+        }
+    }
+    if bid_per_block.is_empty() {
+        return;
+    }
+
+    dump_mir(tcx, false, "lint_tail_expr_drop_order", &0 as _, body, |_, _| Ok(()));
+    let param_env = tcx.param_env(def_id).with_reveal_all_normalized(tcx);
+    let is_closure_like = tcx.is_closure_like(def_id.to_def_id());
+    let move_data = MoveData::gather_moves(body, tcx, param_env, |_| true);
+    let maybe_init = MaybeInitializedPlaces::new(tcx, body, &move_data);
+    let mut maybe_init =
+        maybe_init.into_engine(tcx, body).iterate_to_fixpoint().into_results_cursor(body);
+    for (&block, candidates) in &bid_per_block {
+        let mut all_locals_dropped = ChunkedBitSet::new_empty(move_data.move_paths.len());
+        let mut linted_candidates = BTreeSet::new();
+        let mut drop_span = None;
+        for (idx, &(candidate, place)) in candidates.iter().enumerate() {
+            maybe_init.seek_after_primary_effect(candidate);
+            let MaybeReachable::Reachable(maybe_init_in_future) = maybe_init.get() else {
+                continue;
+            };
+            debug!(maybe_init_in_future = ?maybe_init_in_future.iter().map(|idx| &move_data.move_paths[idx]).collect::<Vec<_>>());
+            let maybe_init_in_future = maybe_init_in_future.clone();
+            for block in drops_reachable_from_location(body, block, place).iter() {
+                let data = &body.basic_blocks[block];
+                if drop_span.is_none() {
+                    drop_span = data
+                        .terminator
+                        .as_ref()
+                        .map(|term| body.source_scopes[term.source_info.scope].span);
+                }
+                debug!(?candidate, "inspect");
+                maybe_init.seek_before_primary_effect(Location {
+                    block,
+                    statement_index: data.statements.len(),
+                });
+                let MaybeReachable::Reachable(maybe_init_now) = maybe_init.get() else { continue };
+                let mut locals_dropped = maybe_init_in_future.clone();
```

Comment:
```suggestion

                // Compute what is dropped along the path from BID to current Drop
                // by comparing what is initialized at the BID and subtracting what is
                // still initialized by the Drop. Anything in between has been dropped
                // along the way.
                let mut locals_dropped = maybe_init_in_future.clone();
```

This code mildly surprised me. I expected us to accumulate the DROP statements that we saw as we traversed the graph. This seems like it could have a "false positive" in that some value might have been *moved* but not *dropped* and thus been removed from the maybe-init set. 

What happens in a case like this?

```rust
struct Droppy { }
impl Drop for Droppy { fn drop(&mut self) { } }
impl Droppy { fn make_temp(&self) { } }

fn send_to_other_thread<T>(t: T) { }

let x = drop(
  (
    {
        Droppy::new().make_temp()
    },
    {
        let v = Droppy::new();
    }
  )
);
```

I believe that the temporary and local are being reordered, however, the local is not initialized at the point where the BID point for the temporary, right?

Reply:
Yes, the lint is missing on this test case. Working on it!

On top of that, the collected drop needs to be filtered through the MaybeInit result because we are working with MIRs at early stage and `drop`s are inserted regardless of its liveness status, ie. they can be moved already as operands.